Add support for literal addition

cdce8p opened the pull request:
Add support for addition of Literal values and str, int and bytes expressions. This is especially useful when working with TypedDicts. Previously this code would have emitted a `literal-required` error.

```py
from typing import Literal, TypedDict

class LookupDict(TypedDict):
    top_var: str
    bottom_var: str
    var: str

def func(d: LookupDict, pos: Literal["top_", "bottom_", ""]) -> str:
    return d[pos + "var"]
```

JukkaL commented:
Can you look at the new errors reported by mypy_primer? If we infer literal types more eagerly, they could generate new false positives.

cdce8p commented:
> Can you look at the new errors reported by mypy_primer?

The primer results look quite promising IMO. For `starlette` the issue is something like this:
```py
from typing import Literal

class SessionMiddleware:
    def __init__(
        self,
        same_site: Literal["lax", "strict", "none"] = "lax",
        https_only: bool = False,
    ) -> None:
        ...
        self.security_flags = "httponly; samesite=" + same_site
        if https_only:  # Secure flag can be used with HTTPS only
            self.security_flags += "; secure"  # <-- error
```
It can be resolved by annotation `self.security_flags` as `str`.

--
For `AutoSplit` the issue looks like this
```py
class CaptureMethodBase:
    description = ""

class BitBltCaptureMethod(CaptureMethodBase):
    description = (
        "\nThe best option when compatible. But it cannot properly record "
        + "\nOpenGL, Hardware Accelerated or Exclusive Fullscreen windows. "
        + "\nThe smaller the selected region, the more efficient it is. "
    )

class ForceFullContentRenderingCaptureMethod(BitBltCaptureMethod):
    description = (
        "\nUses BitBlt behind the scene, but passes a special flag "
        + "\nto PrintWindow to force rendering the entire desktop. "
        + "\nAbout 10-15x slower than BitBlt based on original window size "
        + "\nand can mess up some applications' rendering pipelines. "
    )
```
Just removing the `+` symbols and using implicit string concatenation would resolve the issue here.

--

Not completely sure whats going on with `trio` why the type ignore there is no longer required. However, it's in a test file.
https://github.com/python-trio/trio/blob/3d592af32b4a6dae78525fb5103ac2e58020774a/src/trio/_tests/test_fakenet.py#L185-L189

JukkaL commented:
> Just removing the + symbols and using implicit string concatenation would resolve the issue here.

I think implicit and explicit string concatenation should behave the same. Otherwise it would seem inconsistent. I also know that some people prefer explicit string concatenation as a style thing, since it's easy to accidentally use implicit string concatenation (e.g. omit a comma at the end).

JukkaL commented:
What about inferring a literal addition only if one of the operands is not a literal expression? We have tried to adhere to an informal policy where literal types are not inferred for variables "out of thin air", i.e. an explicit `Literal[...]`, `Final` or similar needs to be somewhere.

JukkaL commented:
In particular, this feels inconsistent to me:
```py
class A:
    a = ""

class B(A):
    a = "a" + "b"

class C:
    a = "a" + "b"

reveal_type(A.a)  # Revealed type is "builtins.str"
reveal_type(B.a)  # Revealed type is "Literal['ab']"  (why literal here?)
reveal_type(C.a)  # Revealed type is "builtins.str"
```

cdce8p commented:
> What about inferring a literal addition only if one of the operands is not a literal expression? We have tried to adhere to an informal policy where literal types are not inferred for variables "out of thin air", i.e. an explicit `Literal[...]`, `Final` or similar needs to be somewhere.

Makes sense. I missed the `"a" + "b"` case when I was refining the implementation. Should be fixed with the latest commits.

brianschubert commented:
Related issue:  #11616

JukkaL commented on mypy/checkexpr.py:
```diff
@@ -3505,6 +3508,63 @@ def visit_op_expr(self, e: OpExpr) -> Type:
+        if isinstance(expr, BytesExpr):
+            return [expr.value], "builtins.bytes", False
+
+        typ = typ or self.accept(expr)
```
Calling `accept` here can be inefficient, since the caller may also call `accept` on the expression at some point, causing duplicate work. This could cause exponential time complexity in nested expression trees (we already have examples of this and it can cause extremely slow type checking). Maybe the caller can pass the expression type, so that we only call `accept` once per expression? Maybe the new logic should be called from `check_op`, when we have the right operand type?

JukkaL commented on mypy/checkexpr.py:
```diff
@@ -3505,6 +3508,63 @@ def visit_op_expr(self, e: OpExpr) -> Type:
+
+        values: list[int | str] = sorted(
+            {
+                val[0] + val[1]  # type: ignore[operator]
```
Can you avoid the `type: ignore` here? In the future we will probably disallow many ignores in code compiled with mypyc, since some type ignores can cause erratic behavior, such as C compilation errors.

cdce8p replied:
> Can you avoid the `type: ignore` here?

Not in an elegant way. The `values` could be either `str` or `int` (for int and bytes). With the comparison of the fallback types we know that left and right values have the same type but not which one. If necessary, it would be possible to split these into two cases with `cast` and set comprehensions for each one. Don't think that it would be an improvement though.

JukkaL commented on test-data/unit/check-literal.test:
```diff
@@ -2960,3 +2960,90 @@ class C(B[Literal["word"]]):
 reveal_type(C().word)  # N: Revealed type is "Literal['word']"
 [builtins fixtures/tuple.pyi]
+
+[case testLiteralAddition]
```
Can you add the interesting bits of the inheritance example I gave in an earlier comment?

JukkaL commented on test-data/unit/check-literal.test:
```diff
@@ -1416,19 +1416,19 @@ c: Literal[4]
+reveal_type('bar' + d)  # N: Revealed type is "Literal['barfoo']"
 
 reveal_type(a.__add__(b))  # N: Revealed type is "builtins.int"
 reveal_type(b.__add__(a))  # N: Revealed type is "builtins.int"
```
Check `a.__add__(a)`. It would be nice if it was consistent with `a + a`, though it's probably not important.

cdce8p replied:
The result will be `int` even if `a: Literal[3]`. The narrowing is only applied for the `a + a` case currently. The method call to `__add__` is handled by `self.check_op` as the fallback case which I haven't modified here.

JukkaL commented on test-data/unit/check-literal.test:
```diff
@@ -2960,3 +2960,90 @@ class C(B[Literal["word"]]):
+bytes_a: Literal[b"a"]
+bytes_b: Literal[b"b"]
+bytes_union_1: Literal[b"a", b"b"]
+bytes_union_2: Literal[b"d", b"c"]
```
Test union with literal type and non-literal type -- e.g. user-defined class that defines `__add__` that accepts `str`. Also test with union of literal type and `Any`.

cdce8p replied:
The narrowing is only applied if all Union items are Literals itself. Thus the inference for these cases didn't change.

Test cases added in https://github.com/python/mypy/pull/18004/commits/36f5e2ead9b02bd6922f88b1761fb70470978011

JukkaL reviewed (commented):
Did a quick pass, mainly to look at the big picture. I can also test this manually later.

brianschubert commented:
Related issue: #11990

Are there any guards against combinatorial explosion? E.g.
```python
from typing import Literal

type HexDigit = Literal["0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "A", "B", "C", "D", "E", "F"]

def foo(a: HexDigit, b: HexDigit, c: HexDigit) -> None:
    reveal_type(a + b + c)
```

cdce8p commented:
> Are there any guards against combinatorial explosion?

That's a good point. Thanks for bringing it to my attention! Implemented a guard in https://github.com/python/mypy/pull/18004/commits/24e2f210221d845a0643f0e48d8f2b7eec01b83c. For now, I've set it to a max of `15` literal values. That should be on the safe side. We could adjust that either way in the future if necessary.

cdce8p commented:
Rebased and added one more commit (https://github.com/python/mypy/pull/18004/commits/c22d5632ea5a1c4841c36c35547c03cfa67cd699) to remove the type ignore.
